package/MDAnalysis/core/universe.py

```diff
@@ -1012,9 +1017,10 @@ def _fragdict(self):
         frags = tuple([AtomGroup(np.sort(ix), self) for ix in frag_indices])
 
         fragdict = {}
-        for f in frags:
+        fraginfo = collections.namedtuple('fraginfo', 'ix, fragment')
+        for i, f in enumerate(frags):
             for a in f:
-                fragdict[a] = f
+                fragdict[a.ix] = fraginfo(i, f)
```

Comment:
Atoms should already just be hashing via their `.ix` (and Universe)

Reply:
`Atom` is derived from `ComponentBase`, so only their `ix` (not their `universe`) is used for hasing.
The reason why I decided to change the `_fragdict` keys to directly use atom indices instead of atoms is performance:

Test system: AtomGroup consisting of 500 fragments with 25 atoms per fragment

Old timings for first access / subsequent access:
* `ag.fragments`: 195 ms / 75 ms
* `ag.fragindices`: 140 ms / 50 ms

New timings for first access / subsequent access:
* `ag.fragments`: 140 ms / 20 ms  --> speedup 1.39 / 3.75
* `ag.fragindices`: 130 ms / 2.7 ms --> speedup 1.08 / 18.5

Since `Universe._fragdict` is cashed and usually (if not always) static, this makes computations involving fragments and fragindices *a lot* faster.

Reply:
Hm, that's odd.  So `Atom.__hash__` is just this:

https://github.com/MDAnalysis/mdanalysis/blob/develop/package/MDAnalysis/core/groups.py#L3028

Is the extra hash call causing the slowdown?

Reply:
I guess so... in my test system that saves 50,000 function calls (`Atom.__hash__` and `hash(self.ix)` for each `Atom`), and these are expensive in Python. Not sure if it's the only reason, though.

Reply:
I think I got it now. The former routine to construct `AtomGroup.fragments` used to be this:
```Python
return tuple(sorted(
            set(a.fragment for a in self),
key=lambda x: x[0].index
))
```
The implicit loop `a.fragment for a in self` needs to create an `Atom` instance for each atom in the group, followed by a call to `Atom.fragment`.

That is much more expensive than the new routine, where the implicit loop is
```
fragdict[aix].fragment for aix in self.ix
```
This only requires instantiating the `self.ix` numpy array, and its members are primitive types.

Reply:
Ok cool, thanks for the performance tweak!